File: nightgrass/src/main/java/com/goodfeel/nightgrass/serviceImpl/OrderService.java
package com.goodfeel.nightgrass.serviceImpl;

import com.goodfeel.nightgrass.data.Order;
import com.goodfeel.nightgrass.data.OrderItem;
import com.goodfeel.nightgrass.dto.OrderDto;
import com.goodfeel.nightgrass.dto.OrderItemDto;
import com.goodfeel.nightgrass.repo.OrderItemRepository;
import com.goodfeel.nightgrass.repo.OrderRepository;
import com.goodfeel.nightgrass.service.IOrderService;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
public class OrderService implements IOrderService {

    private final OrderRepository orderRepository;
    private final OrderItemRepository orderItemRepository;

    public OrderService(OrderRepository orderRepository, OrderItemRepository orderItemRepository) {
        this.orderRepository = orderRepository;
        this.orderItemRepository = orderItemRepository;
    }

    // Retrieve a single order by its ID
    @Override
    public Mono<OrderDto> getOrderById(Long orderId) {
        return orderRepository.findById(orderId)
                .map(this::mapToOrderDto);
    }

    private OrderDto mapToOrderDto(Order order) {
        OrderDto orderDto = new OrderDto();
        orderDto.setOrderId(order.getOrderId());
        orderDto.setOrderNo(order.getOrderNo());
        orderDto.setOrderTotal(order.getTotal());
        return orderDto;
    }

    // Retrieve all items associated with a specific order ID
    @Override
    public Flux<OrderItemDto> getOrderItemsByOrderId(Long orderId) {
        return orderItemRepository.findByOrderId(orderId)
                .map(this::mapToOrderItemDto);
    }

    @Override
    public Mono<Order> findOrderById(Long orderId) {
        return orderRepository.findById(orderId);

    }

    @Override
    public Mono<Order> updateOrder(Order order) {
        return orderRepository.save(order);
    }

    private OrderItemDto mapToOrderItemDto(OrderItem orderItem) {
        OrderItemDto orderItemDto = new OrderItemDto();
        orderItemDto.setProductName(orderItem.getProductName());
        orderItemDto.setImageUrl(orderItem.getImageUrl());
        orderItemDto.setQuantity(orderItem.getQuantity());
        orderItemDto.setProperties(orderItem.getProperties());
        orderItemDto.setUnitPrice(orderItem.getUnitPrice());
        return orderItemDto;
    }

}

File: nightgrass/src/main/kotlin/com/goodfeel/nightgrass/service/OrderService.kt
package com.goodfeel.nightgrass.service

import com.goodfeel.nightgrass.data.Order
import com.goodfeel.nightgrass.data.OrderItem
import com.goodfeel.nightgrass.dto.OrderDto
import com.goodfeel.nightgrass.dto.OrderItemDto
import com.goodfeel.nightgrass.repo.OrderItemRepository
import com.goodfeel.nightgrass.repo.OrderRepository
import com.goodfeel.nightgrass.util.OrderStatus
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Service
class OrderService(
    private val orderRepository: OrderRepository,
    private val orderItemRepository: OrderItemRepository
) : IOrderService {

    private val logger = LoggerFactory.getLogger(OrderService::class.java)


    // Retrieve a single order by its ID
    override fun getOrderById(orderId: Long): Mono<OrderDto> {
        return orderRepository.findById(orderId)
            .map { order ->
                this.mapToOrderDto(order)
            }
    }

    private fun mapToOrderDto(order: Order) = OrderDto(
        orderId = order.orderId!!,
        orderNo = order.orderNo,
        orderTotal = order.total,
        userId = order.userId,
        deliveryAddress = order.deliveryAddress,
        createdAt = order.createdAt,
        orderStatus = order.orderStatus
    )

    // Retrieve all items associated with a specific order ID
    override fun getOrderItemsByOrderId(orderId: Long): Flux<OrderItemDto> {
        return orderItemRepository.findByOrderId(orderId)
            .map { orderItem: OrderItem -> this.mapToOrderItemDto(orderItem) }
    }

    override fun findOrderById(orderId: Long): Mono<Order> {
        return orderRepository.findById(orderId)
    }

    override fun updateOrder(order: Order): Mono<Order> {
        return orderRepository.save(order)
    }

    override fun getOrderByOrderStatus(orderStatus: OrderStatus): Flux<Order> =
        orderRepository.findByOrderStatus(orderStatus)

    override fun findOrderByUserId(userId: String): Flux<Order> {
        logger.debug("Find user's order from DB: $userId")
        return orderRepository.findByUserId(userId)
    }

    override fun mergeOrder(userId: String, guestId: String): Mono<Void> {
        return orderRepository.findByUserId(guestId)
            .flatMap { order ->
                order.userId = userId
                orderRepository.save(order)
            }.then()
    }

    override fun save(orderItem: OrderItem): Mono<OrderItem> {
        return orderItemRepository.save(orderItem)
    }

    private fun mapToOrderItemDto(orderItem: OrderItem) = OrderItemDto(
        productName = orderItem.productName,
        imageUrl = orderItem.imageUrl,
        quantity = orderItem.quantity,
        properties = orderItem.properties,
        unitPrice = orderItem.unitPrice
    )

}
